In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv("../data/application_train.csv")

print(df.shape)

(307511, 122)


In [3]:
X = df.drop("TARGET", axis=1)
y = df["TARGET"]

print("X shape:", X.shape)
print("y shape:", y.shape)

X shape: (307511, 121)
y shape: (307511,)


In [4]:
X = X.drop("SK_ID_CURR", axis=1)

print(X.shape)

(307511, 120)


In [5]:
categorical_cols = X.select_dtypes(
    include=["object", "string", "category"]
).columns

numerical_cols = X.select_dtypes(
    include=["int64", "float64"]
).columns

print("Categorical:", len(categorical_cols))
print("Numerical:", len(numerical_cols))

Categorical: 16
Numerical: 104


In [6]:
X[categorical_cols] = X[categorical_cols].fillna("Unknown")

In [7]:
X[categorical_cols].isnull().sum().sum()

np.int64(0)

In [8]:
X[numerical_cols] = X[numerical_cols].fillna(
    X[numerical_cols].median()
)

In [9]:
X[numerical_cols].isnull().sum().sum()

np.int64(0)

## Data Preprocessing

### 1. Target Separation

The target variable `TARGET` was separated from the input features. `X` contains the applicant information used for prediction, while `y` contains the loan payment outcome.

### 2. Removal of Identifier

The `SK_ID_CURR` column was removed because it is a unique applicant identifier and does not provide meaningful information for predicting credit risk.

### 3. Categorical Feature Handling

The dataset contains 16 categorical features. Missing values in categorical features were replaced with `"Unknown"` rather than assigning an arbitrary existing category.

### 4. Numerical Feature Handling

Missing values in numerical features were replaced using the **median** of the respective feature. Median imputation was selected because it is less sensitive to extreme values and outliers than mean imputation.

### Preprocessing Status

- Target variable separated
- Identifier removed
- Categorical missing values handled
- Numerical missing values handled
- Dataset is ready for categorical encoding and train-test splitting

In [10]:
X = pd.get_dummies(
    X,
    columns=categorical_cols,
    drop_first=True
)

In [11]:
print(X.shape)

(307511, 234)


In [12]:
from sklearn.model_selection import train_test_split

In [13]:
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

In [14]:
print("X_train:", X_train.shape)
print("X_test:", X_test.shape)
print("y_train:", y_train.shape)
print("y_test:", y_test.shape)

X_train: (246008, 234)
X_test: (61503, 234)
y_train: (246008,)
y_test: (61503,)


## Train-Test Split

The preprocessed dataset was divided into training and testing sets using an 80:20 split. The training set is used to train the machine learning models, while the testing set is used to evaluate their performance on unseen data.

Stratification was applied to maintain the original distribution of the target variable in both sets. This is important because the dataset contains an imbalanced target variable.

### Split Summary

- Training samples: **246,008**
- Testing samples: **61,503**
- Features after encoding: **234**
- Training proportion: **80%**
- Testing proportion: **20%**
- Stratification: **Applied**